calicoctl/commands/convert.go

```diff
@@ -0,0 +1,148 @@
+
+Examples:
+  # Create a policy using the data in policy.yaml.
+  calicoctl convert -f ./policy.yaml -o yaml
+
+  # Create a policy based on the JSON passed into stdin.
+  cat policy.json | calicoctl convert -f -
+
+Options:
+  -h --help                     Show this screen.
+  -f --filename=<FILENAME>      Filename to use to create the resource.  If set to
+                                "-" loads from stdin.
+  -o --output=<OUTPUT FORMAT>   Output format. One of: yaml or json.
+                                [Default: yaml]
+
+
+Description:
+  Convert config files between different API versions. Both YAML and JSON formats are accepted.
+
+  The default output will be printed to stdout in YAML format.
+`
+	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	if err != nil {
+		fmt.Printf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.\n", strings.Join(args, " "))
+		os.Exit(1)
+	}
+	if len(parsedArgs) == 0 {
+		return
+	}
+
+	var rp resourcePrinter
+	output := parsedArgs["--output"].(string)
+	// Only supported output formats are yaml (default) and json.
+	switch output {
+	case "yaml", "yml":
+		rp = resourcePrinterYAML{}
+	case "json":
+		rp = resourcePrinterJSON{}
+	default:
+		rp = nil
+	}
+
+	if rp == nil {
+		fmt.Printf("unrecognized output format '%s'\n", output)
+		os.Exit(1)
+	}
+
+	filename := argutils.ArgStringOrBlank(parsedArgs, "--filename")
+
+	// Load the V1 resource from file and convert to a slice
+	// of resources for easier handling.
+	resV1, err := v1resourceloader.CreateResourcesFromFile(filename)
+	if err != nil {
+		fmt.Printf("Failed to execute command: %v\n", err)
+		os.Exit(1)
+	}
+
+	var results []runtime.Object
+	for _, v1Resource := range resV1 {
+		v3Resource, err := convertResource(v1Resource)
+		if err != nil {
+			fmt.Printf("Failed to execute command: %v\n", err)
+			os.Exit(1)
+		}
+		results = append(results, v3Resource)
+	}
+
+	log.Infof("results: %+v", results)
+
+	err = rp.print(nil, results)
+	if err != nil {
+		fmt.Println(err)
+	}
+}
+
+func convertResource(v1resource unversioned.Resource) (runtime.Object, error) {
+	switch strings.ToLower(v1resource.GetTypeMetadata().Kind) {
+	case "node":
+		return convert(conversion.Node{}, v1resource)
+	case "hostendpoint":
+		return convert(conversion.HostEndpoint{}, v1resource)
+	case "workloadendpoint":
+		return convert(conversion.WorkloadEndpoint{}, v1resource)
+	case "profile":
+		return convert(conversion.Profile{}, v1resource)
+	case "policy":
+		return convert(conversion.Policy{}, v1resource)
+	case "ippool":
+		return convert(conversion.IPPool{}, v1resource)
+	case "bgppeer":
+		return convert(conversion.BGPPeer{}, v1resource)
+
+	default:
+		return nil, fmt.Errorf("conversion for the resource type '%s' is not supported", v1resource.GetTypeMetadata().Kind)
+	}
+}
+
+func convert(convRes conversion.Converter, v1resource unversioned.Resource) (conversion.Resource, error) {
```

Comment:
missing a fn description.

Reply:
Not sure you gain much from this extra method, other than making it difficult to explain the difference between convertResource() and convert().

I'm not too fussed if this stays as it is, but another way to structure this would be to have the resource-specific switch statement select a converter to use and then just join two methods together, or have a method that returns the converter for a specific resource type (which is easier to explain than the subtle different between convertResource and convert).

Anyways, it's a nit and I'm talking more than I should.